### ISSUE TITLE
Add support for building Apple Silicon

### ISSUE THREAD

**Issue by MonkeybreadSoftware** [openssl/openssl#12254]
Hi,

you may have watched the Keynote that Apple moves to their own ARM CPUs.

We added this section to 10-main.conf in our copy:

    "darwin64-arm64-cc" => {
        inherit_from     => [ "darwin-common", asm("aarch64_asm") ],
        CFLAGS           => add("-Wall"),
        cflags           => add("-arch arm64 -mmacosx-version-min=11.0 -isysroot /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk"),
        lib_cppflags     => add("-DL_ENDIAN"),
        bn_ops           => "SIXTY_FOUR_BIT_LONG",
        perlasm_scheme   => "macosx",
    },

and we configure with this line:

./Configure shared enable-rc5 zlib darwin64-arm64-cc no-asm

Builds here, although we have to use no-asm as using the assembler didn't build.
Maybe someone can verify this and add it to the next 1.1.x patch?

**Comment by stuartcarnie** [comment]
I submitted PR #12369, which I have built on a DTK. It also includes optimized assembler by specifying the `ios64` asm scheme.

**Comment by mattcaswell** [comment]
Support has been added to master. Not backported to 1.1.1 so far.

**Comment by zowers** [comment]
> Support has been added to master. Not backported to 1.1.1 so far.

@mattcaswell I don't see the merge in commits https://github.com/openssl/openssl/commits/OpenSSL_1_1_1-stable

Also - is backport to 1.0 planned?

**Comment by mattcaswell** [comment]
Like I said support has been added to master but hasn't been backported to 1.1.1 so far.

There definitely won't be a backport to 1.0.x.

**Comment by richsalz** [comment]
Hm, how does adding this compare to the not-adding z/OS support?

**Comment by mattcaswell** [comment]
> Hm, how does adding this compare to the not-adding z/OS support?

I really don't know. I'm not sure whether we should backport this to 1.1.1 or not.

**Comment by paulidale** [comment]
Do we consider a missing platform to be a bug?
It seems like a stretch of the definition.
It would be unlikely to break anything.

**Comment by mattcaswell** [comment]
If we were to hold another vote on this I suspect that the answer would be that this does not count as a bug and therefore not eligible for backport under the current policy.  As with the last time we had this debate there are good arguments to be made as to why we should change our policy - but as of yet I've not seen a concrete proposal for what a new policy might look like.

**Comment by joshuawarner32** [comment]
> It would be unlikely to break anything.
Not sure I agree. It does also block porting of potentially a lot of software to the new platform (speaking as owner of such a project, faced with upgrading openssl). That said, I don't yet know how painful upgrading to the latest will be versus backporting this patch ourselves. YMMV.

**Comment by MonkeybreadSoftware** [comment]
We manually apply the changes needed as we build for Mac ARM already.

I think this should go into one 1.1.1 release soon, independent of any policy.
Once Mac with ARM ships, all projects without support will soon be flooded with people asking for it.

**Comment by shepting** [comment]
Any updates on getting this into either 1.1.1 or a new release? I'm hoping to get `rbenv` working on the DTK and it seems to be blocked on this issue.

**Comment by kaduk** [comment]
> Any updates on getting this into either 1.1.1 or a new release? I'm hoping to get `rbenv` working on the DTK and it seems to be blocked on this issue.

The discussion on #12369 is probably more up-to-date.  (But the short answer seems to be "no updates".)

**Comment by tawmoto** [comment]
Btw, is this command needed for `cflags`?
`-mmacosx-version-min=11.0 -isysroot /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk`

It seems to work also without `version-min` for other libs. Also if Apple changes the sdk path, especially `Xcode-beta.app` dir, it won't work correctly anymore.
Apple recommends adding this to cflags `-target arm64-apple-macos11`
Thanks

**Comment by felixbuenemann** [comment]
I have backported darwin64-arm64-cc support from master in #13476.

**Comment by jgstew** [comment]
Does anyone know what the oldest OpenSSL is to build on ARM Mac? Specifically a FIPS compliant build of OpenSSL on ARM Mac?

### Related to building in general:

- https://blog.andrewmadsen.com/2020/06/22/building-openssl-for.html
- https://cutecoder.org/programming/compile-open-ssl-apple-silicon/

**Comment by zowers** [comment]
@jgstew 1.0 can be built with a patch
there is no FIPS compliant build, the OpenSSL is not validated for the platform

**Comment by felixbuenemann** [comment]
Not sure who asked about it originally, but here is a patch to build OpenSSL 1.0.2u `darwin64-arm64-cc` target:

https://gist.github.com/5f4dcb30ebb3b86e1302e2ec305bac89

It is completely unoptimised (no arm assembly), but I only needed it to build older Ruby versions for compatibility testing.

**Comment by udiakasnuke** [comment]
Not sure from the above comments, so i am asking here again. Is support for arm64(Apple Silicon) added.. I need to build openssl 1.1.1i for arm64 on an Intel based mac currently..
.
Can someone please tell me the steps ?? Thanks!!

**Comment by mattcaswell** [comment]
Yes, this is available in 1.1.1i and was recently introduced via #12369.

Please open a new issue if you are having problems using this. Closing this issue since this feature is now added.

**Comment by udiakasnuke** [comment]
Thanks @mattcaswell --- It worked for me now, just provided a summary here https://github.com/openssl/openssl/pull/12369#issuecomment-751643658 so that everybody is clear as well.

**Comment by pszyrko** [comment]
> @jgstew 1.0 can be built with a patch
> there is no FIPS compliant build, the OpenSSL is not validated for the platform

Hi @zowers Do you know if there is a FIPS compliant build for Apple Silicon? If not, is there any plan for it? Thanks
